### PR TITLE
Implement

### PR DESCRIPTION
**Pull request by tomrf1**
A lambda for checking the status of the Guardian's Kindle publication.

Tests that:
1. the url for the current issue correctly redirects to an S3 path with today's date
2. the number of articles is above a certain threshold

If the tests pass, an email is sent with the article and image counts.
If the tests fail, an email is sent with the cause.

TODO - add cloudwatch schedule to run every night

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+let moment = require('moment');
+
+export class Config {
+    ManifestURL: string = process.env.ManifestURL;
+    KindleBucket: string = process.env.KindleBucket;
+    Stage: string = process.env.Stage;
+    SourceAddress: string = process.env.SourceAddress;
+    PassTargetAddresses: string[] = process.env.PassTargetAddresses.split(',');
```
Comment (regiskuckaertz) [inline src/config.ts]: no need to trim whitespace?

File: src/config.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+let moment = require('moment');
+
+export class Config {
```
Comment (regiskuckaertz) [inline src/config.ts]: You could probably export a singleton object instead of a class?

File: src/lambda.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        .then(() => getS3Objects(config.KindleBucket, `${config.Stage}/${config.Today}`))
+        .then(validatePublicationInfo)
+        .then(sendSuccessEmail)
+        .catch(sendFailureEmail)
+}
+
+//Returns the redirect url, and also checks that it contains today's date
+let getRedirect = (url: Url): Promise<Url> => {
+    return new Promise((resolve, reject) => {
+        http.get(url, response => {
```
Comment (regiskuckaertz) [inline src/lambda.ts]: Could you use a `head` request instead?

File: src/lambda.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            } else {
+                reject(`Expected status code 302 (moved permanently) for url ${url}, got ${response.statusCode}`)
+            }
+        })
+    })
+};
+
+let testRedirect = (url: Url): Promise<number> => {
+    return new Promise((resolve, reject) => {
+        http.get(url, response => {
```
Comment (regiskuckaertz) [inline src/lambda.ts]: ditto

File: src/lambda.ts (hunk 3 above)
Comment (tomrf1) [reply src/lambda.ts]: 👍 good point

Comment (tomrf1) [reply src/lambda.ts]: wow, making other requests is way less convenient https://nodejs.org/api/http.html#http_http_request_options_callback

Comment (regiskuckaertz) [reply src/lambda.ts]: 😵

File: src/lambda.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+let getS3Objects = (bucket: string, prefix: string): Promise<ListObjectsOutput> => {
+    return s3.listObjects({
+        Bucket: bucket,
+        Prefix: prefix
+    }).promise()
+};
+
+let validatePublicationInfo = (result: ListObjectsOutput): Promise<PublicationInfo> => {
+    let articles: S3.Object[] = result.Contents.filter(object => object.Key.endsWith('.nitf.xml'));
```
Comment (regiskuckaertz) [inline src/lambda.ts]: Looks like you only need the `length`

File: src/lambda.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            articleCount: articles.length,
+            imageCount: images.length
+        })
+    } else {
+        return Promise.reject(`Expected at least ${config.MinimumArticleCount} articles, but there are only ${articles.length}`)
+    }
+};
+
+let sendEmail = (subject: string, body: string, targetAddresses: string[]): Promise<SendEmailResponse> => {
+    let request: SendEmailRequest = {
```
Comment (regiskuckaertz) [inline src/lambda.ts:89]: Using JS APIs is so much easier too 🔥

**Review by regiskuckaertz — APPROVED** [review]
That's really great 💯 Let me know if you need another review after adding the CW thing

File: src/lambda.ts (hunk 6 above)
Comment (tomrf1) [reply src/lambda.ts:89]: absolutely, and you still have type safety

File: src/config.ts (hunk 2 above)
Comment (tomrf1) [reply src/config.ts]: I was hoping for something more scala-like that any of these - https://stackoverflow.com/questions/30174078/how-to-define-singleton-in-typescript

File: src/lambda.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+export async function handler(): Promise<SendEmailResponse> {
+    return getRedirect(config.ManifestURL)
+        .then(testRedirect)
+        .then(() => getS3Objects(config.KindleBucket, `${config.Stage}/${config.Today}`))
+        .then(validatePublicationInfo)
+        .then(sendSuccessEmail)
+        .catch(sendFailureEmail)
+}
+
+let headRequestOptions = (url: URL): RequestOptions => {
```
Comment (regiskuckaertz) [inline src/lambda.ts:32]: 🥇

**Pull request merged**